### PR TITLE
Serde struct

### PR DESCRIPTION
**Pull request by Mehrbod2002**
fixes: #2660 
Suppose missed some struct and should add serde features to toml but with cycle importing I suppose collabrators could help to solve it  .

### CODE REVIEW

**Review by ogoffart — COMMENTED** [review]
Thanks a lot for the PR. This is a great start.

However, there is a bunch of things which doesn't make sense to serialize as is.

Because we probably should maintain compatibility over the serialization so we should be careful about that.

Also could you perhaps add a test. I'd say in api/rs/slint/tests/
The test should serialize and deserialize a slint struct and check they are equal, or something like that.

File: internal/core/model.rs [hunk 1]
```diff
@@ -304,6 +304,7 @@ impl<M: Model> Model for Rc<M> {
 
 /// A [`Model`] backed by a `Vec<T>`
 #[derive(Default)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
```
[the remark this answers is not in the document]
Comment (Mehrbod2002) [reply internal/core/model.rs:307]: Thanks for your declaration .
On structs which serde featured added to them except ModelNotify , doesn't need serializing ?

File: internal/core/string.rs [hunk 2]
```diff
@@ -36,6 +36,7 @@ macro_rules! format {
 ///
 /// `SharedString` implements [`Deref<Target=str>`] so it can be easily passed to any function taking a `&str`.
 /// It also implement `From` such that it an easily be converted to and from the typical rust String type with `.into()`
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
```
[the remark this answers is not in the document]
Comment (Mehrbod2002) [reply internal/core/string.rs]: I added test for sharedVector but not this one .
Suppose there are some sophsticate elements on vector not string

**Comment by Mehrbod2002** [comment]
@ogoffart there are some error on test on doc .
i can't recognize it . can you guide what are those ?

File: internal/core/sharedvector.rs [hunk 3]
```diff
@@ -75,6 +77,7 @@ fn capacity_for_grow(current_cap: usize, required_cap: usize, elem_size: usize)
 }
 
 #[repr(C)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
```
[the remark this answers is not in the document]
Comment (Mehrbod2002) [reply internal/core/sharedvector.rs]: Add iteration serialize , can you review it for `into_iter` ?

File: api/rs/slint/Cargo.toml [hunk 4]
```diff
@@ -134,7 +134,8 @@ pin-weak = { version = "1.1", default-features = false }
 num-traits = { version = "0.2", default-features = false }
 
 log = { version = "0.4.17", optional = true }
-
+serde_json = { version = "1.0.96", optional = true }
```
Comment (ogoffart) [inline api/rs/slint/Cargo.toml:137]: serde_json can be a dev-dependency since it is only used in the test

File: api/rs/slint/Cargo.toml [hunk 5]
```diff
@@ -134,7 +134,8 @@ pin-weak = { version = "1.1", default-features = false }
 num-traits = { version = "0.2", default-features = false }
 
 log = { version = "0.4.17", optional = true }
-
+serde_json = { version = "1.0.96", optional = true } 
+serde = { version = "1.0.163", features = ["derive"], optional = true }
```
Comment (ogoffart) [inline api/rs/slint/Cargo.toml:138]: you should also make a "serde" feature that enables "dep:serde" and also the serde feature in i-slint-core

File: api/rs/slint/tests/simple_macro.rs [hunk 6]
```diff
@@ -15,3 +15,15 @@ fn empty_stuff() {
...
+#[cfg(all(feature = "serde", feature = "serde-json"))]
+use slint::SharedVector;
+
+#[cfg(all(feature = "serde", feature = "serde-json"))]
+#[test]
+fn test_serialize_deserialize() {
+    let v = SharedVector::from([1, 2, 3]);
+    let serialized = serde_json::to_string(&v).unwrap();
+    let deserialized: SharedVector<i32> = serde_json::from_str(&serialized).unwrap();
+    assert_eq!(v, deserialized);
```
Comment (ogoffart) [inline api/rs/slint/tests/simple_macro.rs:28]: that's great to test the serialization of vector.
We should also test the serialization of struct declared in a slint! macro.

Comment (Mehrbod2002) [reply api/rs/slint/tests/simple_macro.rs:28]: how can we use struct defined in sline! ?

Comment (Mehrbod2002) [reply api/rs/slint/tests/simple_macro.rs:28]: Do you think can you provide it ? I suppose need expertise like you in slint and its syntax

Comment (Mehrbod2002) [reply api/rs/slint/tests/simple_macro.rs:28]: can we do something like this ?
```rust
#[test]
fn test_serialize_deserialize_person() {
    i_slint_backend_testing::init();
    let mut compiler = slint_interpreter::ComponentCompiler::default();
    compiler.set_style("fluent".into());
    let comp_def = spin_on::spin_on(
        compiler.build_from_source(
            r#"
            export struct TestStruct {
                foo: int,
                bar: string,
            }
    
            export component Test {
                in property <TestStruct> data: [
                    {foo: 1, bar: "test"},
                    {foo: 2, bar: "example"},
                ];
            }
            "#
            .into(),
            "".into(),
        ),
    )
    .unwrap();

    let props = comp_def.properties().collect::<Vec<(_, _)>>();
}
```
for get struct ?

File: api/rs/slint/Cargo.toml [hunk 7]
```diff
@@ -18,6 +18,8 @@ rust-version.workspace = true
 path = "lib.rs"
 
 [features]
+serde = ["i-slint-core/serde"]
+serde-enabled = ["dep:serde", "serde"]
```
Comment (ogoffart) [inline api/rs/slint/Cargo.toml:22]: You do not need the serde enabled feature
```suggestion
serde = ["i-slint-core/serde", "dep:serde"]
```

But also, i don't think the slint crate needs to directly depend on serde, so this should either be moved as a dev-dependency, or removed completely.

```suggestion
serde = ["i-slint-core/serde"]
```

**Review by ogoffart — COMMENTED** [review]
I discussed this with @tronical and we think we should do this PR in several pieces.

1. Make our struct such as SharedVector and SharedString as serializable.
   I think what you have there is good: A feature in slint, forwarded in i-slint-core to enable serialization of our structs.

 2. Then we can enable the serialization of struct declared in slint.  But the question is whether this should be enabled by default as that may actually be a breaking change.
  Maybe it should be opt-in per struct, or maybe it should be done only using import from Rust (#1726) 

So now i think we should try to get the first part polished and merged.

File: api/rs/slint/Cargo.toml (hunk 7 above)
Comment (Mehrbod2002) [reply api/rs/slint/Cargo.toml]: I suppose it should removed completely

Comment (Mehrbod2002) [reply api/rs/slint/Cargo.toml]: @ogoffart there are some error on test on uefi-demo and mcu .
i can't recognize it . can you guide what are those ?

Comment (Mehrbod2002) [reply api/rs/slint/Cargo.toml]: I suppose it will be great if we have optional for serde_json and use as a flag there

**Comment by Mehrbod2002** [comment]
> So just to be clear, i think first we should only keep the changes in the internal/core library  (and add a test there) and also a way to enable the feature from slint/Cargo.toml

Do we miss any of these ? What exactly should we do?
Feature won't call with serde as there are optional?

**Comment by ogoffart** [comment]
The problem is that if we add in the generated code something like
```rust
#[cfg_attr(feature="serde", derive(Serialize, Deserialize))]
struct Foo { 
 ...
}
```

the `feature="serde"` that this represent, is the feature enabled on the app itself, not the slint feature. And it only compiles if the serde feature is enabled in i-slint-core because every field must be serializable.
And right now, some crate might have a serde feature without wanting this option
But things like `image` or `[int]`  are not de-serializable. We'd need to have it opt-in.

Some idea include:

- An an option in slint-build API to enable that.  and a feature in slint-macro crate.

- add something like (in .slint)
```
@rust-attr(cfg_attr(feature="serde", derive(Serialize, Deserialize)))
struct Foo { str: string }
```

- Or something like (also in .slint)
```
@pragma(serde)
struct Foo { str: string }
```

- Or something like #1726 in which you declare the structure in Rust code with `#[derive(...)]` and import it from Slint

So this turns out to be harder than first thought.

But for sure the first step is to allow SharedString and SharedVector to be serialized.
Maybe Image can also be made serializable. Model might be a bit difficult.
We could then generate the derive only for struct that can be serialized.

**Comment by Mehrbod2002** [comment]
@ogoffart 
Could we write an implementation for serialize and deserialize for all possibilities and enable serde as default ? Or should we add some flag due to enable and disable serde ?
I am going to write implementation for serde for Image , is there anything else which contains data ?

**Comment by Mehrbod2002** [comment]
@ogoffart 
I meant , can you do it on this PR ? 
I don't exactly how can I feature it on toml and event worst I don't which toml should I put features table for serde :)

File: api/rs/slint/Cargo.toml [hunk 8]
```diff
@@ -134,6 +134,8 @@ pin-weak = { version = "1.1", default-features = false }
 num-traits = { version = "0.2", default-features = false }
 
 log = { version = "0.4.17", optional = true }
+serde = { version = "1.0.163", features = ["derive"], optional = true }
+serde_json = { version = "1.0.96", optional = true }
```
Comment (ogoffart) [inline api/rs/slint/Cargo.toml:138]: this is also not needed.

**Review by ogoffart — COMMENTED** [review]
In the api/rs/slint/Cargo.toml , just add a feature to forward the core serde feature

```
## Enable the serialization and deserialization of Slint's core types by implementing 
## the traits from the [serde](https://crates.io/crates/serde) crate
serde = ["i-slint-core/serde"]
```

File: api/rs/slint/Cargo.toml [hunk 9]
```diff
@@ -121,7 +121,7 @@ renderer-winit-skia-vulkan = ["i-slint-backend-selector/renderer-winit-skia-vulk
 renderer-winit-software = ["i-slint-backend-selector/renderer-winit-software"]
 
 [dependencies]
-i-slint-core = { version = "=1.0.3", path = "../../../internal/core", default-features = false }
+i-slint-core = { version = "=1.0.3", path = "../../../internal/core", default-features = false, features = ["serde"] }
```
[the remark this answers is not in the document]
Comment (Mehrbod2002) [reply api/rs/slint/Cargo.toml]: done :)

File: internal/compiler/generator/rust.rs [hunk 10]
```diff
@@ -426,6 +426,7 @@ fn generate_struct(name: &str, fields: &BTreeMap<String, Type>) -> TokenStream {
         fields.iter().map(|(name, ty)| (ident(name), rust_primitive_type(ty).unwrap())).unzip();
 
     quote! {
+        #[cfg_attr(feature="serde", derive(Serialize, Deserialize))]
```
Comment (ogoffart) [inline internal/compiler/generator/rust.rs:429]: you can also remove this

File: api/rs/slint/tests/simple_macro.rs [hunk 11]
```diff
@@ -15,3 +15,36 @@ fn empty_stuff() {
...
+}
+
+#[cfg(all(feature = "serde", feature = "serde-json"))]
+#[test]
+fn test_serialize_deserialize_person() {
```
Comment (ogoffart) [inline api/rs/slint/tests/simple_macro.rs:33]: This is not ready, we can remove this test and instead add tests in sharedvector.rs and string.rs 
serde_json can be made a dev dependency of i-slint-core for the test.

Comment (Mehrbod2002) [reply api/rs/slint/tests/simple_macro.rs]: SharedVector has a test above .
Added tests for ShareString

**Comment by Mehrbod2002** [comment]
@ogoffart is there any issue with it?

File: api/rs/slint/Cargo.toml [hunk 12]
```diff
@@ -18,6 +18,9 @@ rust-version.workspace = true
 path = "lib.rs"
 
 [features]
+## Enable the serialization and deserialization of Slint's core types by implementing 
+## the traits from the [serde](https://crates.io/crates/serde) crate
+serde = ["i-slint-core/serde"]
```
[the remark this answers is not in the document]
Comment (Mehrbod2002) [reply api/rs/slint/Cargo.toml]: If move the tests we won't need it any more .

File: api/rs/slint/tests/simple_macro.rs [hunk 13]
```diff
@@ -15,3 +15,24 @@ fn empty_stuff() {
     slint!(export struct Hei { abcd: bool });
     slint!(export global G { });
 }
+
+#[cfg(all(feature = "serde", feature = "serde-json"))]
+use slint::SharedVector;
+
+#[cfg(all(feature = "serde", feature = "serde-json"))]
+#[test]
+fn test_serialize_deserialize_sharedvector() {
```
[the remark this answers is not in the document]
Comment (Mehrbod2002) [reply api/rs/slint/tests/simple_macro.rs]: @ogoffart  let's focus on discuss about way or feature on slint then we can enable this feature

**Comment by Mehrbod2002** [comment]
@ogoffart done ?

**Comment by Mehrbod2002** [comment]
@ogoffart it can't be merge ?

**Comment by tronical** [comment]
Yes, there are merge conflicts. Would you like me to rebase and merge it for you?

**Comment by Mehrbod2002** [comment]
@tronical  Sure , Thanks :)

**Review by tronical — APPROVED** [review]

**Comment by tronical** [comment]
Yes, let's get this in. If you feel like digging further, one thing that would be nice is if the de-serialization into a SharedVector would be possible without the allocation of an intermediate Vec.

**Comment by Mehrbod2002** [comment]
@tronical I will do it these days , almost working to add serde feature in slint .

**Pull request merged**